FirebaseDatabase/Tests/Integration/FIRDatabaseTests.m

```diff
@@ -469,6 +464,65 @@ - (void)testCallbackQueue {
   [database goOffline];
 }
 
+- (void)testSetEmulatorSettingsCreatesEmulatedReferences {
+  id app = [[FIRFakeApp alloc] initWithName:@"testSetEmulatorSettingsCreatesEmulatedReferences"
+                                        URL:self.databaseURL];
+  FIRDatabase *database = [FIRDatabase databaseForApp:app];
+
+  [database useEmulatorWithHost:@"localhost" port:1111];
+  NSString *concatenatedHost = @"localhost:1111";
+
+  FIRDatabaseReference *reference = [database reference];
+
+  NSString *referenceURLString = reference.URL;
+
+  XCTAssert([referenceURLString containsString:concatenatedHost]);
+}
+
+- (void)testSetEmulatorSettingsThrowsAfterRepoInit {
+  id app = [[FIRFakeApp alloc] initWithName:@"testSetEmulatorSettingsThrowsAfterRepoInit"
+                                        URL:self.databaseURL];
+  FIRDatabase *database = [FIRDatabase databaseForApp:app];
+
+  [database reference];  // initialize database repo
+
+  // Emulator can't be set after initialization of the database's repo.
+  XCTAssertThrows([database useEmulatorWithHost:@"a" port:1]);
+}
+
+- (void)testEmulatedDatabaseValidatesOnlyNonCustomURLs {
+  // Set a non-custom databaseURL
+  NSString *databaseURL = @"https://test.example.com";
+  id app = [[FIRFakeApp alloc] initWithName:@"testEmulatedDatabaseValidatesNonCustomURLs0"
+                                        URL:databaseURL];
+  FIRDatabase *database = [FIRDatabase databaseForApp:app];
+
+  // Reference should be retrievable without an exception being raised
+  NSString *referenceURLString = [databaseURL stringByAppendingString:@"/path"];
+  FIRDatabaseReference *reference = [database referenceFromURL:referenceURLString];
+  XCTAssertNotNil(reference);
+
+  app = [[FIRFakeApp alloc] initWithName:@"testEmulatedDatabaseValidatesNonCustomURLs1"
+                                     URL:databaseURL];
+  database = [FIRDatabase databaseForApp:app];
+  [database useEmulatorWithHost:@"localhost" port:1111];
+
+  // Expect production url creates a valid (emulated) reference.
+  reference = [database referenceFromURL:referenceURLString];
+  XCTAssertNotNil(reference);
+  XCTAssert([reference.URL containsString:@"localhost:1111"]);
+
+  // Test emulated url
+  referenceURLString = @"http://localhost:1111/path";
+  reference = [database referenceFromURL:referenceURLString];
+  XCTAssertNotNil(reference);
+  XCTAssert([reference.URL containsString:@"localhost:1111"]);
+
+  // Test non-custom url with different host throws exception
+  referenceURLString = @"https://test.firebaseio.com/path";
+  XCTAssertThrows([database referenceFromURL:referenceURLString]);
```

Comment:
Sorry, I am confused why this test would still throw here. Can you explain?

Reply:
The URL is not a custom host, so it gets validated, but it's invalid because it's a different host than the expected host. This should match the behavior here: https://cs.opensource.google/firebase-sdk/firebase-js-sdk/+/master:packages/database/src/api/Database.ts;l=160?q=database.ts

Reply:
Sorry, I thought it was the other way around. LG.